RCNN_tf.py

```diff
@@ -0,0 +1,338 @@
+        name='rcl'
+):
+    conv_filter_forward_shape = [filter_shape[0], filter_shape[1], num_input_channels, num_filter]
+    conv_filter_recurrent_shape = [filter_shape[0], filter_shape[1], num_filter, num_filter]
+    recurrent_cells_shape = [1, input_data.shape[1], input_data.shape[2], num_filter]
+
+    cell_states = tf.Variable(
+        tf.zeros(recurrent_cells_shape, dtype=PRECISION_TF),
+        trainable=False,
+        name=name + '_cells'
+    )
+
+    forward_weights = tf.Variable(
+        tf.truncated_normal(conv_filter_forward_shape, stddev=std, dtype=PRECISION_TF),
+        trainable=True,
+        name=name + '_forward'
+    )
+
+    recurrent_weights = tf.Variable(
+        tf.truncated_normal(conv_filter_recurrent_shape, stddev=std, dtype=PRECISION_TF),
+        trainable=True,
+        name=name + '_recurrent'
+    )
+
+    bias = tf.Variable(
+        tf.truncated_normal([num_filter], dtype=PRECISION_TF),
+        trainable=True,
+        name=name + '_bias'
+    )
+
+    forward_output = tf.nn.conv2d(input_data, forward_weights, [1, 1, 1, 1], padding='SAME')
+
+    output = None
+    for _ in range(depth):
+        recurrent_output = tf.nn.conv2d(cell_states, recurrent_weights, [1, 1, 1, 1], padding='SAME')
+
+        output = tf.add(forward_output, recurrent_output)
+        output += bias
+
+        cell_states = tf.nn.local_response_normalization(
+            tf.maximum(np.asarray(0.0).astype(PRECISION_NP), output),
+            depth_radius=normalization_feature_maps,
+            bias=1,
+            alpha=alpha,
+            beta=beta,
+            name=name + '_lrn'
+        )
+
+    return output
+
+
+def convolutional_layer(
+        input_data,
+        filter_shape,
+        num_input_channels,
+        num_filter,
+        std=.03,
+        stride=(1, 1),
+        padding='same',
+        name='conv'
+):
+    conv_filter_shape = [filter_shape[0], filter_shape[1], num_input_channels, num_filter]
+    strides = [1, stride[0], stride[1], 1]
+
+    weights = tf.Variable(
+        tf.truncated_normal(conv_filter_shape, stddev=std, dtype=PRECISION_TF),
+        trainable=True,
+        name=name + '_weights'
+    )
+
+    if padding.upper() not in PADDING_LIST:
+        raise ValueError('Padding value is not understood')
+
+    output = tf.nn.conv2d(input_data, weights, strides=strides, padding=padding.upper())
+
+    return output
+
+
+def pooling_layer(
+        input_data,
+        pool_shape,
+        stride=(1, 1),
+        padding='same'
+):
+    ksize = [1, pool_shape[0], pool_shape[1], 1]
+    strides = [1, stride[0], stride[1], 1]
+
+    if padding.upper() not in PADDING_LIST:
+        raise ValueError('Padding value is not understood')
+
+    output = tf.nn.max_pool(input_data, ksize=ksize, strides=strides, padding=padding.upper())
+    return output
+
+
+def global_max_pooling_layer(
+        input_data
+):
+    return tf.reduce_max(input_data, axis=[1, 2])
+
+
+def softmax_layer(
```

Comment:
again, what is the advantage/reason over `tf.nn.softmax`?

Reply:
Again, `tf.nn.softmax` is used. But we need to compute the linear transformation first.